docs/proposals/drafts/20211007-pluggable-authenticators.md

```diff
@@ -0,0 +1,146 @@
+
+We propose to refactor the current way of initializing the authenticators to introduce a common interface, which means all new authenticators have the same basic functions where the adopters need to implement for their custom authenticators. In addition, in this proposal, we aim to provide a self-registering mechanism, so that introducing any new authenticator does not necessary require any additional patch/change in the existing codebase, instead, only the plugin's go source file is needed other than patching the other source files to add the plugin. For that, we propose to use importing for side-effect mechanism, where a protected map of authenticator type and their factories can be used for such scenario.
+
+## Implementation Options
+
+1. Importing for side-effect
+2. Go Plugin Module
+3. Oauth2 Providers
+
+## 1. Importing for side-effects
+
+Like [Option 3](#3-oauth2-providers) except that we do not have a static list of supported authenticators. Instead, each authenticator can register themselves through a map of authenticator types and their factories.
+
+To onboard a new authenticator, the following is needed:
+
+1. Implement the corresponding interface.
+2. Implement a factory.
+3. Register the factory in the authenticators' factories map, this is the part where it registers itself.
+
+The factory should receive three parameters: a config map of strings and interfaces, tenant's name, and a logger instance.
+
+With this approach, an example of tenant configuration would like:
+
+```
+- name: test-oidc
+  id: 1610b0c3-c509-4592-a256-a1871353dbfa
+  authenticator:
+    type: oidc
+    config:
+      clientID: test
+      clientSecret: xyz
+      issuerCAPath: ./tmp/certs/ca.pem
+      issuerURL: http://127.0.0.1:5556/dex
+      redirectURL: https://localhost:8443/oidc/test-oidc/callback
+      usernameClaim: email
+  rateLimits:
+    - endpoint: "/api/metrics/v1/.+/api/v1/receive"
+      limit: 100
+      window: 1s
+    - endpoint: "/api/logs/v1/.*"
+      limit: 100
+      window: 1s
+```
+
+### 2. Go Plugin Module
+
+In this approach, a generic interface for Authentication providers (Authenticators) is introduced. Authentication providers then implement that interface in a go plugin. This plugin then will be built, and the compiled plugin is loaded at run time by Observatorium API, how and where to load the plugin module is defined by the tenants’ config file.
+
+This approach is easy to implement, very flexible in terms of adding more Authenticators without any additional implementation and integration complexity. However, there are few concerns when it comes to trust and security, since Observatorium API would consume a pre-built plugin binary as an authenticator.
```

Comment:
Let's be consistent: `authenticator` should either be used as a proper noun everywhere or nowhere, but not mixed. Let's lower case here.
```suggestion
This approach is easy to implement, very flexible in terms of adding more authenticators without any additional implementation and integration complexity. However, there are few concerns when it comes to trust and security, since Observatorium API would consume a pre-built plugin binary as an authenticator.
```

Reply:
```suggestion
This approach is easy to implement, very flexible in terms of adding more Authenticators without any additional implementation and integration complexity. However, there are few concerns when it comes to trust and security, since the Observatorium API would consume a pre-built plugin binary as an authenticator.
```